crates/bevy_render/src/lib.rs

```diff
@@ -236,3 +242,48 @@ fn check_for_render_resource_context(context: Option<Res<Box<dyn RenderResourceC
         );
     }
 }
+
+fn window_icon_changed(
+    mut map: Local<HashMap<WindowId, Handle<Texture>>>,
+    textures: Res<Assets<Texture>>,
+    mut windows: ResMut<Windows>,
+    asset_server: Res<AssetServer>,
+) {
+    for window in windows.iter_mut() {
+        /* Insert new icon changed */
+        if let Some(WindowIcon::Path(path)) = window.icon() {
+            if let Some(handle) = map.get(&window.id()) {
+                if let Some(handle_path) = asset_server.get_handle_path(handle) {
+                    if handle_path.path() != path {
+                        map.insert(window.id(), asset_server.load(path.clone()));
+                    }
+                }
+            } else {
+                map.insert(window.id(), asset_server.load(path.clone()));
+            }
+        }
+
+        /* Poll load state of handle and set the icon */
+        if let Some(handle) = map.get(&window.id()) {
+            match asset_server.get_load_state(handle) {
+                LoadState::Loaded => {
+                    if let Some(texture) = textures.get(handle) {
+                        /* TODO: Why can I not just unwrap here? */
```

Comment:
I think you can unwrap here?
If the asset is marked as `Loaded` it should exist in the relative `Assets` collection.

Reply:
About the unwrap -> this leads to a panic, but only if you set it at runtime. Not sure why. 
Maybe someone with more experience with bevy_asset can comment on that.

Reply:
The `AssetServer` has be found to have non-intuitive behavior sometimes so this might be one of those weird edge cases.

Reply:
This is probably related to #2224

Reply:
> About the unwrap -> this leads to a panic, but only if you set it at runtime. Not sure why.
Maybe someone with more experience with bevy_asset can comment on that.

Could you try to make a minimal reproduction of this issues? It would probably help us clarify #2224

Reply:
I tried to recreate it here https://github.com/KirmesBude/bevy/tree/repro/issue-2224-asset-loaded-inconsistency, but was unable to do so. 
Looks like there are many different variables at work, but I have not given up yet.

Reply:
Actually it does not seem to occur anymore with a rebase (and my reproduce attempt was on a later commit), 
Maybe I can find out which commit fixed it.